### DISCUSSION TITLE
Port for other languages like C++, C# or JavaScript

### DISCUSSION THREAD

**Discussion by ZhaoXiangXML** [PyO3/pyo3#2452] [Ideas]
I think it's theoretically possible to port PyO3 to support other object-oriented languages like C++ (with smart pointers), C# or JavaScript.

PyO3 has done a great job on exporting Rust to Python with memory management, subclassing and other great stuff. I think these features can be shared with other OO languages so people can write modules in Rust and use them in like C++.

I understand it's a large project. I'd like to know if I want to start on this path where should I start.

**Comment by aganders3** [comment]
Maybe https://github.com/dtolnay/cxx is what you're looking for, at least for C++?

**Reply by davidhewitt** [reply]
Agree with this. PyO3 is very specialised for Python bindings and I don't think you'll find much joy from trying to generalise it to other languages. Better to use cxx or other projects tailored for that specific language.

**Reply by ZhaoXiangXML** [reply]
It's just really annoying to implement bindings for different languages while they have so much in common.

**Reply by messense** [reply]
Have you considered https://github.com/Dushistov/flapigen-rs?

**Reply by ZhaoXiangXML** [reply]
> Have you considered https://github.com/Dushistov/flapigen-rs?

Thank you, that looks easier to port. I'll take further look into it.
